### ISSUE TITLE
[Impeller] Poor performance with large nonzero fill paths.

### ISSUE THREAD

**Issue by Mairramer** [flutter/flutter#135458]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

I am using syncfusion_flutter_charts with  ColorFiltered. The performance is so slow.
[App example](https://github.com/Mairramer/impeller_example/tree/master)

1. Run app by default using impeller
2. then disable the impeller 

### Expected results

Good or excellent performance.

### Actual results

Performance is poor.

### Code sample

<details><summary>Code sample</summary>

```dart
ColorFiltered(
          colorFilter: AppThemeProvider.currentTheme.name == AppThemesMode.dark.name
              ? const ColorFilter.mode(Colors.white, BlendMode.color)
              : const ColorFilter.mode(Colors.transparent, BlendMode.color),
          child: SfSignaturePad(
            key: _signatureGlobalKey,
            backgroundColor: Colors.transparent,
            strokeColor: Colors.black,
            minimumStrokeWidth: 3.0,
            maximumStrokeWidth: 4.0,
            onDraw: (offset, time) {
              setState(() => currentOffset = offset);
            },
          ),
        ),
```
</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/50643541/573efb92-e7b2-4131-a124-0af548bf3ee3

https://github.com/flutter/flutter/assets/50643541/0b16e806-c9c2-4bad-bde6-568f27d272a4

</details>

### Logs

<details><summary>Logs</summary>

```console
[Failed to change device orientation: Error Domain=UIScene ErrorDomain Code=101 "None of the requested orientations are supported by the view controller. Requested: landscapeRight; Supported: portrait" UserInfo={NSLocalizedDescription=None of the requested orientations are supported by the view controller. Requested: landscapeRight; Supported: portrait}](https://www.google.com/search?sca_esv=568251480&q=Failed+to+change+device+orientation:+Error+Domain%3DUIScene+ErrorDomain+Code%3D101+%22None+of+the+requested+orientations+are+supported+by+the+view+controller.+Requested:+landscapeRight;+Supported:+portrait%22+UserInfo%3D%7BNSLocalizedDescription%3DNone+of+the+requested+orientations+are+supported+by+the+view+controller.+Requested:+landscapeRight;+Supported:+portrait%7D&spell=1&sa=X&ved=2ahUKEwikr-7jvMaBAxUOrJUCHUTVCx8QBSgAegQIChAB)
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.13.0, on macOS 13.0 22A380 darwin-arm64, locale pt-BR)
    • Flutter version 3.13.0 on channel stable at /Users/mairramersilva/fvm/versions/3.13.0
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision efbf63d9c6 (6 weeks ago), 2023-08-15 21:05:06 -0500
    • Engine revision 1ac611c64e
    • Dart version 3.1.0
    • DevTools version 2.25.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/mairramersilva/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = /Users/mairramersilva/Library/Android/sdk
    • ANDROID_SDK_ROOT = /Users/mairramersilva/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] IntelliJ IDEA Community Edition (version 2022.1.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.82.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.72.0

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 13.0 22A380 darwin-arm64
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 116.0.5845.187

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by jonahwilliams** [comment]
Happy to take a look at the performance but you need to include a reproducible example, not just a video recording.

**Comment by Mairramer** [comment]
> Happy to take a look at the performance but you need to include a reproducible example, not just a video recording.

@jonahwilliamsI added an example for testing.

**Comment by jonahwilliams** [comment]
This overhead is all tessellation:

![image](https://github.com/flutter/flutter/assets/8975114/9176b237-9868-4564-90d2-7ef88180b11a)

**Comment by jonahwilliams** [comment]
I'm going to merge this into https://github.com/flutter/flutter/issues/134432 because its the same underlying issue

**Comment by dnfield** [comment]
Just as a tip here:

Looking at the videos, the application is very clearly "just" drawing strokes, but the profile that @jonahwilliams shared indicates that they're getting drawn with `PaintingStyle.fill` (otherwise they wouldn't hit that tessellatio ncode).

Tessellating these strokes as fills is _very_ expensive in Impeller, and it would be far better to tessellate them as strokes.

**Comment by jonahwilliams** [comment]
Reopening because this is a bit different: we can actually fix this case by being smarter about large non-zero fill contours.

**Comment by jonahwilliams** [comment]
I have a pending fix in https://github.com/flutter/engine/pull/46282 that makes the performance comparable to Skia.

**Comment by jonahwilliams** [comment]
This change will be reverted so I'm re-opening the issue.

**Comment by jonahwilliams** [comment]
We think this is fixed now (from the linked issue #143077
